### ISSUE TITLE
What should be the internal representation of `DerivationPath`?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3107]
Firstly, I'd love to just not have it since it's the same as slice/vec containing sufficiently specific types. But we probably really need it to implement `FromStr` and `Display`. For `Display` we'd ideally have an unsized version too. (or rather have `DerivationPath` and `DerivationPathBuf`)

But there's one more thing. With the addition of `NormalDerivationPath` there's optimization opportunity to not reallocate when converting between the paths. If `DerivationPath` contained `u32`s that are converted on the edges we could make one conversion zero-cost (just `unsafe` cast the types) and the other would only cost checking the numbers. But to do this we must lose `AsRef<[ChildNumber]>` and `Index` impls. We can still have accessors of individual numbers (just converting) and with the unsized type it shouldn't be a big deal. (But we're doing the exact opposite of what I'd prefer ideally.)

I wonder if it's worth it.

**Comment by apoelstra** [comment]
I think losing `AsRef<[ChildNumber]>` is probably fine. This functionality can mostly be emulated if we have a way to iterate over `DerivationPath`s yielding the various childnumber types, and if we implement `ops::Index<usize>` to return a `ChildNumber` (maybe doing some conversions to do so).

We won't be able to implement `ops::Index<ops::Range*<usize>>` for the same reasons we can't implement `AsRef` but I think this is a pretty rare thing for people to want.

**Comment by Kixunil** [comment]
> if we implement `ops::Index<usize>` to return a `ChildNumber` (maybe doing some conversions to do so).

That's impossible because we have to return a reference.

> We won't be able to implement `ops::Index<ops::Range*<usize>>`

We could if we had unsized `DerivationPath`.

**Comment by apoelstra** [comment]
> > if we implement `ops::Index<usize>` to return a `ChildNumber` (maybe doing some conversions to do so).
> 
> That's impossible because we have to return a reference.
>

Damn. I should've checked that before posting. Frustrating.

We can provide our own `.get` or `.nth` function I guess.

> > We won't be able to implement `ops::Index<ops::Range*<usize>>`
> 
> We could if we had unsized `DerivationPath`.

If we had unsized `DerivationPath` and if whatever childnumber types we might want to return were `#[repr(transparent)]` wrappers on `usize`? I like this idea but it's a pretty strong API limitation. I think it's fine for our childnumbers to be repr(transparent) but making `DerivationPath` unsized also means that it basically can't have any non-zero-sized fields execept the unsized one. So we can't put any extra metadata or anything into it. Maybe that's ok.

(Technically Rust allows structs with arbitrary fields to be unsized; you just make the last field be `[u8]` or `dyn Trait` or whatever. But good luck constructing such a thing. AFAICT you can't do it without wading waaay into the weeds of manual memory allocation and messing with `Layout` and so on. I tried this in rust-simplicity with our `Value` type recently and couldn't do it.)

**Comment by Kixunil** [comment]
> We can provide our own `.get` or `.nth` function I guess.

Yes. Also maybe one day `Index` will support non-references since we have GAT today.

> whatever childnumber types we might want to return were `#[repr(transparent)]` wrappers on `usize`?

`u32` rather.

> making `DerivationPath` unsized also means that it basically can't have any non-zero-sized fields execept the unsized one

I'd want to have `*Buf` types too. (so what is currently `DerivationPath` would become `DerivationPathBuf` - exactly what we did with `Script`.)

> But good luck constructing such a thing. AFAICT you can't do it without wading waaay into the weeds of manual memory allocation and messing with `Layout` and so on.

Actually it can be very simple if you know the length statically - just create a generic struct, initialize it with sized array and then castt it. But that's not possible for custom types. IIRC I've seen some advanced hacks around this that I was considering for `ControlBlock`.

**Comment by apoelstra** [comment]
All sounds great to me. And looking forward to seeing any advanced hacks you can find in the future.

**Comment by jaoleal** [comment]
>  If DerivationPath contained u32s that are converted on the edges we could make one conversion zero-cost (just unsafe cast the types) and the other would only cost checking the numbers. But to do this we must lose AsRef<[ChildNumber]> and Index impls.

I think its a great trade-off, i would argue that access to internals, besides being unsafe, certainly pass an idea that the developer using the lib might need it, we are not offering all the methods needed to use the implementation. If any feature, that we dont have, needs some access to internals, sure thing that we will be notified with a opportunity to improve the lib
